### PR TITLE
add display names to components

### LINKED ISSUE #1080: Improve Components DisplayNames properties
**Is your feature request related to a problem? Please describe.**
Hi there,

sometimes I find it extremely difficult to properly debug base components in `React Dev Tools` due to lack of `Display Name` (I think this is it?). For smaller layouts it will not be an issue, however sometimes we can construct more complex ones using only components from theme-ui. Debugging those can be a little bit of pain as some sample React's component tree can look like:

image

**Describe the solution you'd like**
A little bit cleaner react's component tree to debug.

### PR DESCRIPTION
**Pull request by dburles**
React devtools and Emotion will now render the correct component names, for https://github.com/system-ui/theme-ui/issues/811

### CODE REVIEW

**Comment by lachlanjc** [comment]
This is awesome! Any reason we can’t do it with [the simpler syntax](https://github.com/rebassjs/grid/blob/master/src/index.js#L25) though?

**Comment by dburles** [comment]
For that instance it is due to the emotion `styled` API not reading it for whatever reason

**Comment by dburles** [comment]
The function name approach is needed when using `forwardRef` https://reactjs.org/docs/forwarding-refs.html

**Comment by jxnblk** [comment]
Rad, thanks! Would changing the Box component to not use the `styled` HOC affect this at all? (See https://github.com/system-ui/theme-ui/pull/692#issuecomment-600131921)

**Comment by dburles** [comment]
No it should be okay, if it's just a regular component, it will only need to pass a named function to `forwardRef`. These changes will definitely conflict with #692. I think if this were merged first, it shouldn't be very difficult to reconcile.

**Comment by yuriybelike** [comment]
Hey! Thanks for this PR related to my issue.

I may not be approaching this the right way, but when I inspect the components in the [Deploy preview](https://deploy-preview-817--theme-ui.netlify.com/) I can only see `-Box` prefixes, instead of `-Image`, `-Heading` and such.

image

**Comment by dburles** [comment]
> Hey! Thanks for this PR related to my issue.
> 
> I may not be approaching this the right way, but when I inspect the components in the [Deploy preview](https://deploy-preview-817--theme-ui.netlify.com/) all I see are `-Box` suffixes, instead of `-Image`, `-Heading` and such.
> 
> image

Yeah that is odd, I'll have a look into it

**Comment by dburles** [comment]
Ok after digging into it a bit, I think the best solution to ensure emotion renders the correct names is to add the label property to all components. I'll push a fix soon.

**Comment by dburles** [comment]
> Rad, thanks! Would changing the Box component to not use the `styled` HOC affect this at all? (See [#692 (comment)](https://github.com/system-ui/theme-ui/pull/692#issuecomment-600131921))

I'm kinda second guessing using the emotion `label` option as they're not stripped in production, though it is one solution. I've gone down a bit of a rabbit hole and the indirection through the `Box` component is causing problems in that emotion always resolves to `Box` unless we otherwise specify a `label` on every component.

This should be able to be greatly simplified by writing components without reference to `Box` such as:

```js
/** @jsx jsx */
import React from 'react'
import { jsx, useThemeUI } from 'theme-ui'

export const Alert = React.forwardRef(function Alert(props, ref) {
  const { theme } = useThemeUI()
  return (
    <div
      ref={ref}
      {...props}
      sx={{
        display: 'flex',
        alignItems: 'center',
        px: 3,
        py: 2,
        fontWeight: 'bold',
        color: 'white',
        bg: 'primary',
        borderRadius: 4,
        variant: 'alerts.' + props.variant
      }}
    />
  )
})
``` 

Some of these components use `styled-system` style props, which is another reason why they all inherit from the `Box` component, e.g.:

```js
<Alert variant='secondary' mb={2}>Secondary</Alert>
```

IMO, there is no need for this now that we have the `sx` prop, it is odd to have two API's and only adds confusion. This might also simplify the TS refactor over at #692. Also, this definitely plays into the `variant` prop idea also: #800.

**Comment by dburles** [comment]
I think I'll roll with the `label` option for now as that'll ensure that emotion renders the correct names. I'll try have that up soon.

**Comment by dburles** [comment]
Alright so I think this is ready! Emotion will now render the component names, it's unfortunate but they'll all be named like `css-191ogd4-Box-Text` for example (always with `Box`). So to summarise this PR, it includes two things:

1. React devtools will now render the component names
2. Emotion will generate classnames containing the component names

It would be even better if we could get CSS sourcemaps working too, the babel plugin will require changes to support theme-ui. Something to look at in future :)

File: packages/components/src/Alert.js [hunk 1]
```diff
@@ -1,20 +1,23 @@
...
-      px: 3,
-      py: 2,
-      fontWeight: 'bold',
-      color: 'white',
-      bg: 'primary',
-      borderRadius: 4,
-    }}
-  />
-))
+export const Alert = React.forwardRef(function Alert(props, ref) {
```
Comment (jxnblk) [inline packages/components/src/Alert.js:4]: For my understanding: is the named `function` syntax still necessary with the Emotion `label` property added below?

Comment (jxnblk) [reply packages/components/src/Alert.js:4]: Also, holy smokes! I don't think I realized the `label` property would work more generally in the `sx` prop – that's probably worth putting in a guide in the docs or mentioning somewhere 🤔

Comment (dburles) [reply packages/components/src/Alert.js:4]: > For my understanding: is the named `function` syntax still necessary with the Emotion `label` property added below?

Just for React devtools

Comment (jxnblk) [reply packages/components/src/Alert.js:4]: Does the `displayName` static property no longer work for React devtools?

Comment (dburles) [reply packages/components/src/Alert.js:4]: It does but just not when you’re using `forwardRef` https://reactjs.org/docs/forwarding-refs.html#displaying-a-custom-name-in-devtools

Comment (dburles) [reply packages/components/src/Alert.js:4]: Just thought I'd double check it, it does indeed work with `forwardRef`! Strange, I am sure that it wasn't working at the time, anyway, I'd probably stick with this approach as it's what's recommended and is more concise.

**Comment by atanasster** [comment]
Hi @dburles - this is a great PR. 

Can you please upgrade it to resolve conflicts and we also have some new components
- Paragraph
- Switch

Also whats the final status on displayName? 

I hope we can merge this soon, as @lachlanjc is working on a rewrite to remove emotion/styled and it seems there is some overlap here.

**Comment by dburles** [comment]
@atanasster Thanks. I've just fixed the merge conflicts. Named functions serve the same purpose as `displayName`, so either way it doesn't make any difference. 

I think we should probably remove the `label` property:

1. It isn't stripped in production, which means additional data over the wire.
2. It's tedious to correctly handle alongside the `__css` prop on `Box`. For instance `CheckBoxChecked` (in `Checkbox`) which renders the `SVG` component has custom styles defined using the `__css` prop. The means in each lower level component (`SVG` in this case), for those styles to function we would have to merge `label` with those styles so that the styles are correctly applied. It's possible to do, but unnecessarily tedious. This is only a problem for such instances (which I think are few, but still...).

I'm not sure `label` is worth the additional complexity, at least with the current implementation with everything using `Box`.

If we can make a decision on that I am happy to remove it and at least sort out the naming in devtools for now. Emotion `label` can always be revisited later.

**Comment by atanasster** [comment]
Thanks @dburles thats great

I would say remove label. You are correct this PR should be only about the devTools and we can decide about label in some other PR if its requested by users.

Would you mind switching to static displayName, its a bit more standard and will be also easier for the rework that @lachlanjc is doing.

**Comment by dburles** [comment]
I can't find why `displayName` appears to be working now ([it didn't used to](https://github.com/facebook/react/issues/13703)), the [React docs](https://reactjs.org/docs/forwarding-refs.html#displaying-a-custom-name-in-devtools) still recommend naming the functions instead of using `displayName` when using `forwardRef`. So I'm not sure which is incorrect.

**Comment by atanasster** [comment]
@dburles i think displaying the component name looks cleaner too, no?

**Comment by dburles** [comment]
I mean like, technically you shouldn't be able to use `displayName` on `forwardRef` components. I do recall it not working some time in the past, now for some reason it seems to work. I'm not sure what that means because the docs still recommend providing a named function to forwardRef (and not displayName). Unless I can confirm that `displayName` is now supported on `forwardRef` components I wouldn't recommend it.

**Comment by atanasster** [comment]
I see you are correct it was recommended by bvaughn in that thread. You make the call, seems both work so we should be good.

**Comment by dburles** [comment]
I'll aim to have the label removed by early next week.

**Comment by atanasster** [comment]
Fantastic, really looking forward to have your PR merged.

For displayName - any choice you make is fine by me, but until then (monday) if @hasparus @lachlanjc have any notes about it here

**Review by hasparus — APPROVED** [review]
Great PR, @dburles! Thank you!

@atanasster let's merge this after the `label` removal.

**Comment by atanasster** [comment]
@dburles - would you mind updating Paragraph and Switch, just so we dont forget to bring them up to date as well.

**Comment by dburles** [comment]
@atanasster Whoops, nice spot. Just updated those, should be good to go then.

**Comment by atanasster** [comment]
@dburles - thanks a lot - its ready to merge. Let's just wait @hasparus to confirm if we are merging only bug fixes to release a 0.6.0 stable, or if we are also merging features.

**Review by atanasster — APPROVED** [review]
Thanks a bunch, great PR

**Comment by hasparus** [comment]
> Let's just wait @hasparus to confirm if we are merging only bug fixes to release a 0.6.0 stable, or if we are also merging features.

You wrote it like I'm some sort of authority.

This PR is open since March, and IMHO it's a bugfix, as I'd expect proper names in React DevTools.

**Comment by hasparus** [comment]
Sorry it took so long to merge it, @dburles 🙏

**Pull request merged**